pkg/kapp/crdupgradesafety/change_validator.go

```diff
@@ -241,6 +241,40 @@ func MinimumPropertiesChangeValidation(diff FieldDiff) (bool, error) {
 	}
 }
 
+// DefaultValueChangeValidation adds a validation check to ensure that
+// default values are not changed in a CRD schema:
+// - No new value can be added as default that did not previously have a
+// default value present
+// - Default value of a field cannot be changed
+// - Existing default value for a field cannot be removed
+// This function returns:
+// - A boolean representation of whether or not the change
+// has been fully handled (i.e. the only change was to a field's default value)
+// - An error if either of the above criteria are not met
+func DefaultValueChangeValidation(diff FieldDiff) (bool, error) {
+	handled := func() bool {
+		diff.Old.Default = &v1.JSON{}
+		diff.New.Default = &v1.JSON{}
+		return reflect.DeepEqual(diff.Old, diff.New)
+	}
+
+	switch {
+	case diff.Old.Default == nil && diff.New.Default != nil:
+		return handled(), fmt.Errorf("new value added as default when previously no default value existed: %+v", diff.New.Default.Raw)
+
+	case diff.Old.Default != nil && diff.New.Default == nil:
+		return handled(), fmt.Errorf("default value has been removed when previously a default value existed: %+v", diff.Old.Default.Raw)
+
+	case diff.Old.Default != nil && diff.New.Default != nil:
+		if !reflect.DeepEqual(diff.Old.Default.Raw, diff.New.Default.Raw) {
```

Comment:
`slices.Equal` or `bytes.Equal` probably makes more sense to use here than `reflect.DeepEqual`. Reflection should only be used when necessary. The reason we use it when dealing with the full `v1.JSONSchemaProps` resources is because we need to ensure that the entire object is equivalent as opposed to comparing a specific field (which is significantly more complex to write and generalize for an API that can change)

Reply:
Ah that's good to learn, I've updated this to use `bytes.Equal` in the latest commit.